Scaffold: Upgrade to react-native v0.71.7

danielsousaio opened the pull request:
## Ticket

PLAT-7049

## Changes introduced

- react 18.2/react-native 0.71.7
- typescript support
- node version bumps
- ruby version bumps
- python version bumps
- docker images versions bumps
- documentation updates (quick links section)
- scaffold upgrade script via npx (check how to use it down below on the upgrade docs)
- introduction of changelogs for both scaffold and the repo
- new maintenance helper scripts (check the changelog)
- redesign of README and docs homepage

All of the commits that are related with updating files that were overriding the default template include the output of a git diff - those commits messages start with "update /custom":

```sh
git diff b89bf7efd2818b69673961d87378321cc6e8afc4:/scaffold/template/filepath \
b89bf7efd2818b69673961d87378321cc6e8afc4:/scaffold/template/custom/filepath
```

This explains why any file has been updated and thus is self-documenting.

_Note: Ignore all changes made in the /dist directory, those are automatically generated by the bootstrap script._

## Quick links for reviewing new documentation

- [Simpler README](https://github.com/crowdbotics/modules/tree/upgrade/react-71)
- [Scaffold Changelog](https://github.com/crowdbotics/modules/tree/upgrade/react-71/scaffold/CHANGELOG.md)
- [Changelog](https://github.com/crowdbotics/modules/tree/upgrade/react-71/CHANGELOG.md)
- [New Docs Homepage](https://github.com/crowdbotics/modules/blob/upgrade/react-71/docs/index.md) - this is what users will see in the https://crowdbotics.github.io/modules/ page
- [Upgrading your scaffold](https://github.com/crowdbotics/modules/blob/upgrade/react-71/docs/scaffold-upgrade.md)

The PR template was also simplified reducing the number of questions asked to only two, and links to the relevant documentation included:

- [Modules checklist](https://github.com/crowdbotics/modules/blob/upgrade/react-71/docs/modules-checklist.md) - Help testing that everything in that list actually works is welcomed!
- [Scaffold checklist](https://github.com/crowdbotics/modules/blob/upgrade/react-71/docs/scaffold-checklist.md)

## Test and review

- Checkout this branch `upgrade/react-71`.
- Double check that you are running compatible Node.js, Python and Ruby versions.
- `yarn install`.
- `pipenv install`.
- Perform your testing here.

### Run local emulator

```
yarn run demo
cd demo
npx react-native-start
```

When prompted: choose iOS or android to start the respective emulator.

mabdullahabid reviewed (changes requested):
Unable to install Python 3.8.2 on a Mac M1 using pyenv.

https://github.com/pyenv/pyenv/wiki/Common-build-problems#configure-error-internal-configure-error-for-the-platform-triplet-please-file-a-bug-report-in-macos

> This means that the Python version you're installing doesn't support your MacOS and/or XCode version. In particular:

> XCode 13.3+ is officially supported by CPython since 3.8.13 and 3.9.8. We use downstream patches to support it in some older versions, too, see ["Python versions with extended support" in the README](https://github.com/pyenv/pyenv#python-versions-with-extended-support).
The ARM64 architecture is supported since 3.8.10 and 3.9.1

I was able to install 3.8.13

Can we bump up the Python version to 3.8.13?

mabdullahabid commented:
We should also add ios/Podfile.lock to the scaffold as that is what is required for saving/restoring cache when the Update CocaPods dependencies step is executed!

danielsousaio commented:
> Can we bump up the Python version to 3.8.13?
For sure! Done

danielsousaio commented:
> We should also add `ios/Podfile.lock` to the scaffold as that is what is required for saving/restoring cache when the `Update CocaPods dependencies` step is executed!

Could you please run `pod install` and push the `Pipfile.lock` here? I can't do it on Linux.

mabdullahabid commented:
![Screenshot 2023-05-09 at 12 39 09 PM](https://github.com/crowdbotics/modules/assets/22291306/2672810d-d148-43e0-b2d7-b7da37587594)

So, Daniel, the Podfile.lock file gets recreated everytime pod install is run. This means that the restore and cache never works as the checksum changes everytime pod install is run. I think we need to get those additional files which are created as part of the initial install and save it. I'm gonna try pushing those changes and see if it fixes it.

alaisgomes reviewed (commented):
`App.tsx` is still a thing in this PR. We either need to update Studio to match any acceptable extension for the scaffold files or revert this change. 
I can work on the Studio feature if you want, to support this. Just let me know

danielsousaio commented:
> Edit: nvm I see the script that replaces the custom. All good

Yeah, I noticed that this filename was "hardcoded" in some places and ended up renaming it in the post install script just to make the transition easier.

Hesbon5600 commented on CHANGELOG.md:
```diff
@@ -0,0 +1,34 @@
+
+Introduced upgrade utility for scaffold users - available via `npx crowdbotics/modules`.
+
+Starts an upgrade process that gives the user the option to upgrade his app to any of the scaffold released versions. Validates that the user is currently in a supported version for the option picked.
```
... upgrade `his` app .. to `their`

mabdullahabid commented:
@danielsousaio please resolve conflicts

tan-duong reviewed (dismissed):
LGTM

danielsousaio commented:
Added https://github.com/crowdbotics/modules/pull/875/commits/398db66635424b5aa25fcb43880b578371c68b23 and https://github.com/crowdbotics/modules/pull/875/commits/ac40666a0517695502c127b11ab6056c4b81fa1c and rebased onto master.

GaboGomezT commented on package.json:
```diff
@@ -1,26 +1,31 @@
   "type": "module",
   "engines": {
-    "node": ">14.0"
+    "node": ">16"
```
Is the intention to require versions of at least 17?

danielsousaio replied:
Good catch, thanks. This was meant to require 16 *and* above. Fixed it in 08d8c56.
